### DISCUSSION TITLE
Error: "/bin/sh: 0: Illegal option -o pipefail"

### DISCUSSION THREAD

**Discussion by gudeh** [The-OpenROAD-Project/OpenROAD-flow-scripts#330] [General] [answered]
Hi there, I used the OpenROAD-flow-scripts previously a couple months ago and worked fine from start to end. 

Although, I recently updated to the newest version, left everything as is, but I got this error: 
```
gudeh@snwoflake:~/Desktop/OpenROAD-flow-scripts/flow$ make
[INFO][FLOW] Using platform directory ./platforms/nangate45
/bin/sh: 0: Illegal option -o pipefail
/bin/sh: 0: Illegal option -o pipefail
/bin/sh: 0: Illegal option -o pipefail
/bin/sh: 0: Illegal option -o pipefail
/bin/sh: 0: Illegal option -o pipefail
/bin/sh: 0: Illegal option -o pipefail
/bin/sh: 0: Illegal option -o pipefail
/bin/sh: 0: Illegal option -o pipefail
/bin/sh: 0: Illegal option -o pipefail
```
I noticed that commenting out line 103 in the Makefile there is no error: "#.SHELLFLAGS    = -o pipefail -c".

**Comment by maliberty** [comment, marked as answer]
What shell are you using?

**Reply by rovinski** [reply]
The log indicates `/bin/sh` so the question is why isn't the Makefile picking up the `SHELL`. @gudeh what OS and version are you using?

**Reply by gudeh** [reply]
```
gudeh@snwoflake:~/Desktop/bills$ hostnamectl
   Static hostname: snwoflake
         Icon name: computer-laptop
           Chassis: laptop
        Machine ID: 0d33b2c6ab8644aba892d3dc38c58e86
           Boot ID: 7fcd75b093da4b3c8d3c6a2f65b459a9
  Operating System: Ubuntu 20.04.3 LTS
            Kernel: Linux 5.13.0-27-generic
      Architecture: x86-64
```

**Reply by rovinski** [reply]
Do you have the environment variable `SHELL` set when you invoke make? Or do you manually set it somewhere in the Makefile? That's the only reason I can think of why it wouldn't be set correctly here: https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/blob/9941baeed7b9dc0e3b4fcfe4711dfdbcb499bf0f/flow/Makefile#L102

**Reply by gudeh** [reply]
Yes, shell environment variable seems fine: 
```
gudeh@snwoflake:~$ printenv
SHELL=/bin/bash
```
I did nothing but follow the instructions for local build in the docs and than call `make`, no editions. It is funny though, cause as I said, commenting line 103 the flow runs. Aside from an error message for klayout though, which I suppose it is caused from the same issue:
`Makefile:645: *** KLayout not found in PATH.  Stop.`

Although if I check it on terminal it is installed:
```
gudeh@snwoflake:~/Desktop/OpenROAD-flow-scripts/flow$ klayout -v
KLayout 0.26.2
```

**Reply by rovinski** [reply]
I can't reproduce the issue so I'm not sure I can help. It looks like some kind of environment issue to me. Obviously from the log, `/bin/sh` is being called somewhere instead of `/bin/bash`. You might try changing line 102 to be `SHELL = /bin/bash` and see if that helps. You can also try adding debug prints like `$(info SHELL: $(SHELL))` to try to narrow down the issue.

**Reply by maliberty** [reply]
What does "echo $SHELL" give?

**Reply by gudeh** [reply]
```
gudeh@snwoflake:~/Desktop/OpenROAD-flow-scripts/flow$ echo $SHELL
/bin/bash

```

**Reply by gudeh** [reply]
@rovinski 's tip on rewritting line 102 to `SHELL = /bin/bash`, wihtout the '?' seems to solve the issue. The flow ran without any error or warning!

**Reply by vvbandeira** [reply]
> ```
> gudeh@snwoflake:~/Desktop/bills$ hostnamectl
>    Static hostname: snwoflake
>          Icon name: computer-laptop
>            Chassis: laptop
>         Machine ID: 0d33b2c6ab8644aba892d3dc38c58e86
>            Boot ID: 7fcd75b093da4b3c8d3c6a2f65b459a9
>   Operating System: Ubuntu 20.04.3 LTS
>             Kernel: Linux 5.13.0-27-generic
>       Architecture: x86-64
> ```

@gudeh are you running Ubuntu natively or through Windows WSL? I have seen a similar issue but was on WSL.

**Reply by gudeh** [reply]
I am running on Ubuntu natively, no emulation involved.

**Comment by maliberty** [comment]
would you report the output of "/bin/bash -version"

**Reply by gudeh** [reply]
```
gudeh@snwoflake:~/Desktop$ /bin/bash -version
GNU bash, version 5.0.17(1)-release (x86_64-pc-linux-gnu)
Copyright (C) 2019 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>

This is free software; you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
```

**Comment by maliberty** [comment]
I have no clue either.  I agree with the suggestion:

You can also try adding debug prints like $(info SHELL: $(SHELL)) to try to narrow down the issue.

**Reply by gudeh** [reply]
By adding this line to Makefile it goes: 
```
gudeh@snwoflake:~/Desktop/OpenROAD-flow-scripts/flow$ make
SHELL: /bin/sh
[INFO][FLOW] Using platform directory ./platforms/nangate45
/bin/sh: 0: Illegal option -o pipefail
/bin/sh: 0: Illegal option -o pipefail
/bin/sh: 0: Illegal option -o pipefail
/bin/sh: 0: Illegal option -o pipefail
/bin/sh: 0: Illegal option -o pipefail
/bin/sh: 0: Illegal option -o pipefail
/bin/sh: 0: Illegal option -o pipefail
/bin/sh: 0: Illegal option -o pipefail
/bin/sh: 0: Illegal option -o pipefail
Makefile:647: *** KLayout not found in PATH.  Stop.

```

**Comment by gudeh** [comment]
As I said in the discussion previously, by removing the '?' in line 102 in the Makefile solved the issue. I am sorry for my inexperience but I also have no idea what is going on haha. If you guys want to investigate further the issue to cover a possible bug, just tell me what data I can provide. Thus far the problem seems solved here.

**Reply by maliberty** [reply]
> As I said in the discussion previously, by removing the '?' in line 102 in the Makefile solved the issue. I am sorry for my inexperience but I also have no idea what is going on haha. If you guys want to investigate further the issue to cover a possible bug, just tell me what data I can provide. Thus far the problem seems solved here.

If you are content with that solution I'm ok to close this.

**Reply by simonwaid** [reply]
Same issue here, removing the ? works

**Reply by maliberty** [reply]
@Unkraut3 what platform/OS are you using?  Are you able to debug further as suggested above?

**Reply by rovinski** [reply]
@maliberty I actually ran into this issue on Ubuntu 20.04 because the shell defaulted to `/bin/sh`. Is there a reason to use the environment shell instead of just overriding to bash?

**Reply by maliberty** [reply]
@rovinski https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/commit/545c8d56dfbd047f43b4b7efe6ede38a966ec842

**Reply by rovinski** [reply]
smh

**Comment by maliberty** [comment]
Somehow you are overriding SHELL but from your descriptions I don't see how.  It would either have to be an envar or a command line option afaik.

**Comment by vvbandeira** [comment]
FYI @ahmadelrouby

**Comment by gkamendje** [comment]
I ran into this issue when I updated from Ubuntu 18.04 to 20.04 in WSL. I had to run
 `sudo dpkg-reconfigure dash`
to remove dash as the default shell interpreter. This solved the problem for me.

**Reply by gudeh** [reply]
This didn't solve here. I could only make it work by removing the `?` in line: 
`SHELL          ?= /bin/bash`

**Comment by gudeh** [comment]
Hi everyone. I tried installing OpenROAD in a computer in my university and I once again got in this same error and solved by removing the `?` in line: 
`SHELL ?= /bin/bash`

I also think it may be useful to mention that I had to install some packages manually and the `./etc/DependencyInstaller.sh -dev` didn't seem to solve all the job by itself, as it seemed to be doing some months ago.
